job-runner/chain/chain.go

```diff
@@ -15,19 +15,29 @@ import (
 
 // chain represents a job chain and some meta information about it.
 type chain struct {
+	// A note on the read/write mutexes in chain:
+	//   A lot of the methods in chain call each other - be careful not to make
+	//   nested RLock calls within one goroutine, or a Lock call from another
+	//   goroutine can cause a deadlock (and it will be very annoying to debug).
 	jcMux    *sync.RWMutex
```

Comment:
All the mutexes are confusing. Can a single mutex guard any and all changes to the chain? Or would that serialize access too severely?